Add ability to report start of paragraphs in braille

nvdaes opened the pull request:
<!-- Please read and fill in the following template, for an explanation of the sections see:
https://github.com/nvaccess/nvda/blob/master/projectDocs/dev/githubPullRequestTemplateExplanationAndExamples.md
Please also note that the NVDA project has a Citizen and Contributor Code of Conduct which can be found at https://github.com/nvaccess/nvda/blob/master/CODE_OF_CONDUCT.md. NV Access expects that all contributors and other community members read and abide by the rules set out in this document while participating or contributing to this project. This includes creating or commenting on issues and pull requests. 

Please initially open PRs as a draft.
When you would like a review, mark the PR as "ready for review". 
See https://github.com/nvaccess/nvda/blob/master/.github/CONTRIBUTING.md.
-->

### Link to issue number:
Closes #16895
### Summary of the issue:
NVDA should have an option to indicate paragraphs in braille, so that reading, especially books or long pieces of text, can be more comfortable and meaningful.
### Description of user facing changes
- In the braille settings panel, a new combo box can be used for setting NVDA to show two spaces, a pilcrow or no marker at the start of paragraphs in braille. This requires that the reading by paragraph checkbox is enabled.

### Description of development approach
- A new option for this feature has been added in the braille section of config.conf.
- In config["braille"], a string is assigned to get the startParagraphMarker.
- A getParagraphStartMarker function has been added, so that it can be used in different places of braille.py, and in this way we can also make the PILCROW translatable.
- In the getFormatFieldBraille function of braille.py: when the isAtStart parameter is True, NVDA will check if it's configured to read by paragraph and if it should present the start paragraph indicator, through the getParagraphStartMarker function.
- In the _set_windowEndPos function of braille.py: if the braille wrap word and paragraph start options are enabled, and the start window pos to be set is minor than or equal to the lenghts of paragraph start indicator + 1 (1 corresponding to the text separator), tart pos is set to the start position of the last region if it starts with paragraph start indicator plus text separator. This handles some cases where the paragraph start indicator is not displayed when the braille word wrap is enabled.
- The braille settings subpanel has been updated in the gui.
- A new ParagraphStartMarker class has been added to configFlags.py

### Testing strategy:
Tested manually:
- Reading scrolling foreard and backward, and moving forward and backward through paragraphs, the braille display.
- Moving backward and forward between paragraph with right and left arrows, and control+down and control + up arrow.
- Checking the start position to handle cases where the paragraph start indicator wasn't displayed in braille.
- Tested options added in the braille settings panel.
### Known issues with pull request:
None
### Code Review Checklist:

<!--
This checklist is a reminder of things commonly forgotten in a new PR.
Authors, please do a self-review of this pull-request.
Check items to confirm you have thought about the relevance of the item.
Where items are missing (eg unit / system tests), please explain in the PR.
To check an item `- [ ]` becomes `- [x]`, note spacing.
You can also check the checkboxes after the PR is created.
A detailed explanation of this checklist is available here:
https://github.com/nvaccess/nvda/blob/master/projectDocs/dev/githubPullRequestTemplateExplanationAndExamples.md#code-review-checklist
-->

- [x] Documentation:
  - Change log entry
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] UX of all users considered:
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [x] API is compatible with existing add-ons.
- [x] Security precautions taken.

<!-- Please keep the following -->
<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

## Summary by CodeRabbit

- **New Features**
  - Introduced a configuration option to indicate the start of paragraphs in braille for improved accessibility.
  - Added a feature in the Braille settings panel that allows users to toggle the visibility of paragraph starts.

- **Documentation**
  - Updated user guide to include information on the new Braille settings related to paragraph starts.
  - Enhanced user documentation to reflect the impact of paragraph indicators on reading experience and navigation. 

- **Bug Fixes**
  - Improved logic in braille formatting to better reflect user preferences based on updated configuration settings.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

nvdaes committed:
Add ability to report start of paragraphs in braille

nvdaes committed:
Reformat files

nvdaes committed:
Improve translators comment

nvdaes committed:
Fix help event for show paragrahph start in braille

nvdaes committed:
Reformat file

nvdaes commented:
@coderabbitai review

nvdaes committed:
Improve documentation, suggested by coderabbitai

nvdaes commented:
@coderabbitai review

CyrilleB79 commented on user_docs/en/changes.md:
```diff
@@ -9,6 +9,7 @@
 * Enhanced Microsoft Word comment command: press twice to present comment content in browsable message. (#16800, @Cary-Rowen)
 * Enhanced Microsoft Excel notes command: press twice to present notes content in browsable message. (#16878, @Cary-Rowen)
 * NVDA can now be configured to report font attributes in speech and braille separately. (#16755)
+* NVDA can now be configured to indicate the start of paragraphs in braille when the reading by paragraph option is enabled (#16895,  @nvdaes)
```
* Change the order of phrases in the sentence to improve clarity
* Put the real option name
* Use quoting marks

```suggestion
* When the "Read by paragraph" option is enabled, NVDA can now be configured to indicate the start of paragraphs in braille. (#16895,  @nvdaes)
```

CyrilleB79 commented on user_docs/en/userGuide.md:
```diff
@@ -2163,6 +2163,10 @@ This means that you do not have to scroll the display at the end of each line ev
 This may allow for more fluent reading of large amounts of text.
 It is disabled by default.
 
+##### Show paragraphs start {#BrailleSettingsShowParagraphStart}
```
```suggestion
##### Show paragraphs start {#BrailleShowParagraphStart}
```

CyrilleB79 commented on source/gui/settingsDialogs.py:
```diff
@@ -4363,8 +4363,22 @@ def makeSettings(self, settingsSizer):
+			keyPath=["braille", "showParagraphStart"],
+			conf=config.conf,
+		)
+		self.bindHelpEvent("BrailleSettingsShowParagraphStart", self.showParagraphStartCombo)
```
The most recent usage is to avoid using settings panel name in the anchor name, in case options are moved in the future. For example, with the increasing number of braille options, we may imagine that we will have 2 separate (and more specific) braille settings panels.

```suggestion
		self.bindHelpEvent("BrailleShowParagraphStart", self.showParagraphStartCombo)
```

CyrilleB79 commented on user_docs/en/userGuide.md:
```diff
@@ -2163,6 +2163,10 @@ This means that you do not have to scroll the display at the end of each line ev
 It is disabled by default.
 
+##### Show paragraphs start {#BrailleSettingsShowParagraphStart}
+
```
Could you add a table with the possible options and the default one, as done for any recently introduced feature flag option?

nvdaes replied:
Done.

CyrilleB79 commented on user_docs/en/userGuide.md:
```diff
@@ -2163,6 +2163,10 @@ This means that you do not have to scroll the display at the end of each line ev
 
+##### Show paragraphs start {#BrailleSettingsShowParagraphStart}
+
+- If enabled and "Read by paragraph" is checked, two spaces will be displayed to indicate the start of a paragraph.
```
* remove leading dash
* Link the "Read by paragraph" option

```suggestion
If enabled and "[Read by paragraph](#BrailleSettingsReadByParagraph)" is checked, two spaces will be displayed to indicate the start of a paragraph.
```

Moreover, it would be worth rephrasing this paragraph with a bit more details. But I am not familiar with "Read by paragraph" option and will not be able to test on a braille display before next week, so I won't be able to make a detailed suggestion.

Maybe something like the following (to be adapted and polished):
When "Read by paragraph" option is enabled, we face [issueToBeDescribed, e.g. difficulty to identify paragraph start].
Enabling this option enable to mitigate this issue adding two spaces at the beginning of paragraphs.

nvdaes committed:
Apply suggestions from code review

Co-authored-by: Cyrille Bougot <cyrille.bougot2@laposte.net>

nvdaes committed:
Apply suggestion from code review

nvdaes committed:
Improve user guide, addressing review suggestions

nvdaes commented:
@CyrilleB79 , thanks for your review. I think I've addressed all your suggestions. Let me know if I should apply any other changes.

XLTechie commented on user_docs/en/userGuide.md:
```diff
@@ -2163,6 +2163,15 @@ This means that you do not have to scroll the display at the end of each line ev
+|Default |Disabled|
+
+If enabled and "Read by paragraph" is checked, two spaces will be displayed to indicate the start of a paragraph.
+This can be specially helpful in applications used to read large pieces of text like structured documents or books, where reporting paragraphs start maybe important to understand the structure of the read contents, and even to set bookmarks or annotations based on paragraph position.
```
English smoothing.

```suggestion
This can be especially helpful in applications used to read large pieces of text, like structured documents or books.
In such documents, knowing where paragraphs start maybe important to understand the structure of the content, or to set bookmarks or annotations based on paragraph position.
```

XLTechie commented on user_docs/en/userGuide.md:
```diff
@@ -2163,6 +2163,15 @@ This means that you do not have to scroll the display at the end of each line ev
 This may allow for more fluent reading of large amounts of text.
 It is disabled by default.
 
+##### Show paragraphs start {#BrailleShowParagraphStart}
```
I think the plural of the word "paragraph" is not correct usage here. I'm not 100% sure of that, but reading it in English it just "doesn't sound right".

```suggestion
##### Show paragraph start {#BrailleShowParagraphStart}
```

nvdaes committed:
Apply suggestions from code review

Co-authored-by: Luke Davis <8139760+XLTechie@users.noreply.github.com>

nvdaes commented:
@XLTechie , thanks for your review. I've applied your suggestions. Let me know if I should fix other things.

nvdaes commented:
@coderabbitai review

nvdaes committed:
Use single space as indicator for paragraphs, to get a double space

nvdaes committed:
Revert "Use single space as indicator for paragraphs, to get a double space"

This reverts commit 0d9400b4a882f78d7b6cab667e2baffd29e5e841.

nvdaes committed:
Revert "Revert "Use single space as indicator for paragraphs, to get a double space""

This reverts commit 23c017af83ec16a060c2ed647d5dd2c9bad84725.

nvdaes commented:
These days I'm testing this in my daily reading. I've observed that, once, when scrolling down the braille display, two spaces haven't been shown at the start of the paragrahp and I needed to scroll up and down. This may be related to a separate issue, since sometimes the first character is not shown and scrolling up and down is needed. I thought that trying to set two spaces as the paragraph start indicator may fix this issue, but it doesn't work.
So I think that I won't made more changes for this PR unless required by reviewers.

nvdaes committed:
Prevent that indicator of paragraphs is kipped if the word wrap option is enabled

nvdaes committed:
Reformat files

nvdaes committed:
Set window poswhen paragraph start indicator should start the raw text of last region

nvdaes committed:
Improve setting the start position of window when it should start with paragraph indicator

nvdaes committed:
Handle moving to previous window when word wrap is active

nvdaes committed:
Improvement to prevent skipping paragraph start indicators when word wrap is enabled

nvdaes commented:
@coderabbitai review

nvdaes commented:
@coderabbitai full review

seanbudd commented on source/braille.py:
```diff
@@ -1756,6 +1762,17 @@ def _set_windowEndPos(self, endPos):
 					break
 		except ValueError:
 			pass
+		# Prevent that paragraph start indicator is skipped
```
Can you explain this comment more? what is the case here

nvdaes replied:
When the braille wrap option is enabled, sometimes spaces indicating the start of paragraphs aren't displayed in braille. I'm reading a book in Thorium Reader, a system to read epub and other formats in browse mode, and in specific parts of the book paragraph are displayed without spaces.

seanbudd replied:
can you please expand this comment to explain this case further

seanbudd commented on source/gui/settingsDialogs.py:
```diff
@@ -4484,6 +4499,9 @@ def onTetherToChange(self, evt: wx.CommandEvent) -> None:
 		tetherChoice = [x.value for x in TetherTo][evt.GetSelection()]
 		self.brailleReviewRoutingMovesSystemCaretCombo.Enable(tetherChoice != TetherTo.FOCUS.value)
 
+	def onReadByParagraphChange(self, evt):
```
please add a type for event

nvdaes commented on source/braille.py:
```diff
@@ -330,6 +330,9 @@
 TEXT_SEPARATOR = " "
 
+# Indicator at the start of paragraphs
+PARAGRAPH_START = " "
```
I would like that the paragraph start indicator is customizable. I didn't implement this to start with a simpler gui. We mayuse two spaces by default, and add an edit box to customize this.

nvdaes committed:
Address review: add type for event

nvdaes committed:
Merge remote-tracking branch 'origin/master' into brailleParagraphs

nvdaes committed:
Add ability to customize paragraph start indicator

nvdaes committed:
Add ability to customize paragraph start and update user guide

seanbudd commented:
Rather than a custom field,  this should be one setting with 3 options:
- No paragraph start marker (default)
- Use double space to indicate paragraph start
- Use pilcrow to indicate paragraph start i.e. `¶`. It seems like liblouis supports this symbol

nvdaes committed:
Merge remote-tracking branch 'origin/master' into brailleParagraphs

nvdaes committed:
Expand explanation about a case where paragraph start indicator may not be displayed if it's not handled

nvdaes committed:
Add explanatory comment

nvdaes committed:
Apply format

nvdaes committed:
Add combo box with double space and pillcrow for paragraph start markers

nvdaes committed:
Update documentation

seanbudd commented on source/config/configSpec.py:
```diff
@@ -85,6 +85,8 @@
 		optionsEnum="ReviewRoutingMovesSystemCaretFlag", behaviorOfDefault="NEVER")
 	readByParagraph = boolean(default=false)
+	paragraphStartMarker = option("", " ", "¶", default="")
+	paragraphStart=string(default=" ")
```
```suggestion
```

seanbudd commented on source/config/configFlags.py:
```diff
@@ -251,3 +251,20 @@ def _displayStringLabels(self):
+	def _displayStringLabels(self):
+		return {
+			# Translators: This is a label for a paragraph start marker.
+			self.NONE: _("Don't use paragraph start markers (default)"),
```
```suggestion
			self.NONE: _("No paragraph start marker (default)"),
```

seanbudd commented on source/braille.py:
```diff
@@ -1880,16 +1877,18 @@ def _set_windowEndPos(self, endPos):
 		except ValueError:
 			pass
-		# Prevent that paragraph start indicator is skipped
+		# When word wrap is enabled, the first block of spaces maybe removed from the current window
```
```suggestion
		# When word wrap is enabled, the first block of spaces may be removed from the current window.
```

seanbudd commented on source/braille.py:
```diff
@@ -1880,16 +1877,18 @@ def _set_windowEndPos(self, endPos):
 			pass
-		# Prevent that paragraph start indicator is skipped
+		# When word wrap is enabled, the first block of spaces maybe removed from the current window
+		# This may prevent to display start of paragraphs
+		# Check if this case should be handled
```
```suggestion
		# This may prevent displaying the start of paragraphs.
```

seanbudd commented on source/gui/settingsDialogs.py:
```diff
@@ -4389,18 +4390,18 @@ def makeSettings(self, settingsSizer):
-			keyPath=["braille", "showParagraphStart"],
-			conf=config.conf,
+		# Translators: This is a label for a combo-box in the Braille settings panel to select paragraph start markers.
+		labelText = _("Paragraph start markers:")
```
```suggestion
		labelText = _("Paragraph start marker:")
```

seanbudd commented on user_docs/en/userGuide.md:
```diff
@@ -2163,41 +2163,17 @@ This means that you do not have to scroll the display at the end of each line ev
-If enabled and "Read by paragraph" is checked, two spaces will be displayed to indicate the start of a paragraph.
+If "Read by paragraph" is checked, the selected start marker will be displayed to indicate the start of a paragraph.
 This can be especially helpful in applications used to read large pieces of text, like structured documents or books.
 In such documents, knowing where paragraphs start maybe important to understand the structure of the content, or to set bookmarks or annotations based on paragraph position.
```
```suggestion
In such documents, knowing where paragraphs start may be useful to understand the structure of the content, or to set bookmarks or annotations based on paragraph position.
```

seanbudd commented on user_docs/en/userGuide.md:
```diff
@@ -2163,41 +2163,17 @@ This means that you do not have to scroll the display at the end of each line ev
-If enabled and "Read by paragraph" is checked, two spaces will be displayed to indicate the start of a paragraph.
+If "Read by paragraph" is checked, the selected start marker will be displayed to indicate the start of a paragraph.
 This can be especially helpful in applications used to read large pieces of text, like structured documents or books.
 In such documents, knowing where paragraphs start maybe important to understand the structure of the content, or to set bookmarks or annotations based on paragraph position.
```
>or to set bookmarks or annotations based on paragraph position.

Is this through the text editor UX, or are you referring to add-on functionality?

nvdaes replied:
This is related to applications to read books, for example,Thorium Reader. We can press NVDA+enter to perform a click (activate the current object), sothat the app knows the current element of the DOM (for the current HTML document of the epub being read), and then press control+b toset a bookmark. Then pressing control+n we can open the navigation menu, use the arrow keys to select the bookmarks tab, and create an annotation pressing the edit button, adding several lines of text. If we do this in the middle of a paragraph, the bookmark isadded based on the HTML element, so it won't be very precise. The placeMarkers add-on (maintained by me) considers the number of characters from the start of the text, but bookmarks are added for each document with a different title, so in Thorium I prefer to use the reading app feature, since chapters of books areplaced in separate HTML documents, but the title is for the whole book.

seanbudd commented on user_docs/en/userGuide.md:
```diff
@@ -2163,6 +2163,17 @@ This means that you do not have to scroll the display at the end of each line ev
+
+| . {.hideHeaderRow} |.|
+|---|---|
+|Options |Don't use paragraph start markers, Double space, Pillcrow|
```
```suggestion
|Options |No paragraph start marker, Double space, Pilcrow|
```

seanbudd commented on user_docs/en/userGuide.md:
```diff
@@ -2163,6 +2163,17 @@ This means that you do not have to scroll the display at the end of each line ev
+| . {.hideHeaderRow} |.|
+|---|---|
+|Options |Don't use paragraph start markers, Double space, Pillcrow|
+|Default |Don't use paragraph start markers|
```
```suggestion
|Default |No paragraph start marker|
```

seanbudd commented on user_docs/en/userGuide.md:
```diff
@@ -2163,6 +2163,17 @@ This means that you do not have to scroll the display at the end of each line ev
 
+##### Paragraph start markers {#BrailleParagraphStartMarkers}
+
+| . {.hideHeaderRow} |.|
```
please move this table to the end of the settings group, as is the new standard described in userGuideStandards

nvdaes replied:
Done

seanbudd commented on user_docs/en/userGuide.md:
```diff
@@ -2163,6 +2163,17 @@ This means that you do not have to scroll the display at the end of each line ev
 This may allow for more fluent reading of large amounts of text.
 It is disabled by default.
 
+##### Paragraph start markers {#BrailleParagraphStartMarkers}
```
```suggestion
##### Paragraph start marker {#BrailleParagraphStartMarkers}
```

nvdaes committed:
Apply suggestions from code review

Co-authored-by: Sean Budd <seanbudd123@gmail.com>

nvdaes committed:
Address review: apply standars to user guide

nvdaes committed:
Merge remote-tracking branch 'origin/master' into brailleParagraphs

nvdaes commented:
@seanbudd , I think I've addressed all your suggestions. For some reason, a file from include folder has been modified but not in commits made today from me.
AppVeyor has failed due to time exceeding 60 minutes.

seanbudd commented:
@nvdaes - you must have committed it by mistake, please reset our checkout masters `include/microsoft-ui-uiautomation`

seanbudd committed:
Merge branch 'master' into brailleParagraphs

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

seanbudd reviewed (approved):
Approved, pending include/microsoft-ui-uiautomation getting fixed

nvdaes committed:
Remove accidentally committed file

nvdaes committed:
Reset source/speech/commands.py to master, since it's not related to this PR

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

nvdaes commented:
@seanbudd, I've fixed include/microsoft-ui-uiautomation and also a speech file not related to this PR.

seanbudd commented on source/speech/commands.py:
```diff
@@ -80,7 +80,7 @@ def _getFormattedDevInfo(self):
 		return (
 			f"CancellableSpeech ("
-			f"{ 'cancelled' if self._checkIfCancelled() else 'still valid' }"
+			f"{ 'cancelled' if self._checkIfCancelled() else 'still valid'}"
```
@nvdaes - this will be added until #16866 is merged

nvdaes replied:
OK, then I'll revert e1d6241

seanbudd commented on source/config/configFlags.py:
```diff
@@ -251,3 +251,20 @@ def _displayStringLabels(self):
+			self.NONE: _("No paragraph start marker (default)"),
+			# Translators: This is a label for a paragraph start marker.
+			self.SPACE: _("Double space (  )"),
+			# Translators: This is a label for a paragraph start marker.
```
This might help translators understand the symbol better and know how to translate it.

```suggestion
			# Translators: This is a label for a paragraph start marker.
			# Pilcrow is a symbol also known as "paragraph symbol" or "paragraph marker".
			# Ensure this is consistent with other strings with the context "paragraphMarker".
```

seanbudd commented on source/config/configFlags.py:
```diff
@@ -251,3 +251,20 @@ def _displayStringLabels(self):
+class ParagraphStartMarker(DisplayStringStrEnum):
+	NONE = ""
+	SPACE = " "
+	PILCROW = "¶"
```
Can this also be made translatable?

```suggestion
	# Translators: This is a paragraph start marker used in braille.
	# The default symbol is the pilcrow,
	# a symbol also known as "paragraph symbol" or "paragraph marker".
	# This symbol should translate in braille via LibLouis automatically.
	# If there is a more appropriate character for your locale,
	# consider overwriting this (e.g. the reversed pilcrow ⁋).
	# You can also use Unicode Braille such as ⠘⠏.
	# Ensure this is consistent with other strings with the context "paragraphMarker".
	PILCROW = pgettext("paragraphMarker", "¶")
```

seanbudd commented on user_docs/en/userGuide.md:
```diff
@@ -2163,6 +2163,17 @@ This means that you do not have to scroll the display at the end of each line ev
+
+| . {.hideHeaderRow} |.|
+|---|---|
+|Options |No paragraph start marker, Double space, Pilcrow|
```
```suggestion
|Options |No paragraph start marker, Double space (  ), Pilcrow (¶)|
```

seanbudd commented on source/config/configFlags.py:
```diff
@@ -251,3 +251,20 @@ def _displayStringLabels(self):
+			# Translators: This is a label for a paragraph start marker.
+			self.SPACE: _("Double space (  )"),
+			# Translators: This is a label for a paragraph start marker.
+			self.PILCROW: _("Pilcrow (¶)"),
```
```suggestion
			self.PILCROW: pgettext("paragraphMarker", "Pilcrow (¶)"),
```

seanbudd commented on source/config/configFlags.py:
```diff
@@ -251,3 +251,20 @@ def _displayStringLabels(self):
+			# Translators: This is a label for a paragraph start marker.
+			self.NONE: _("No paragraph start marker (default)"),
+			# Translators: This is a label for a paragraph start marker.
+			self.SPACE: _("Double space (  )"),
```
```suggestion
			self.SPACE: pgettext("paragraphMarker", "Double space (  )"),
```

nvdaes commented:
OK, pre commit bot fixed e1d6241. Then I won't change anything. Thanks @seanbudd for approving this.

seanbudd commented on source/config/configFlags.py:
```diff
@@ -251,3 +251,20 @@ def _displayStringLabels(self):
+	def _displayStringLabels(self):
+		return {
+			# Translators: This is a label for a paragraph start marker.
+			self.NONE: _("No paragraph start marker (default)"),
```
```suggestion
			self.NONE: pgettext("paragraphMarker", "No paragraph start marker (default)"),
```

Qchristensen reviewed (approved):
User guide reads fine.  Is it worth clarifying that the pilcrow character is (what some people may know as) a paragraph marker? I must admit I had encountered the character before but never knew its name myself until now

seanbudd commented on source/config/configSpec.py:
```diff
@@ -85,6 +85,7 @@
 	reviewRoutingMovesSystemCaret = featureFlag(\
 		optionsEnum="ReviewRoutingMovesSystemCaretFlag", behaviorOfDefault="NEVER")
 	readByParagraph = boolean(default=false)
+	paragraphStartMarker = option("", " ", "¶", default="")
```
can you please make this an int to correspond to the index of the option?

nvdaes replied:
Can you provide more details about this, please?

seanbudd commented on user_docs/en/userGuide.md:
```diff
@@ -2163,6 +2163,17 @@ This means that you do not have to scroll the display at the end of each line ev
+
+If "Read by paragraph" is checked, the selected start marker will be displayed to indicate the start of a paragraph.
+This can be especially helpful in applications used to read large pieces of text, like structured documents or books.
+In such documents, knowing where paragraphs start may be useful to understand the structure of the content, or to set bookmarks or annotations based on paragraph position.
```
```suggestion
In such documents, knowing where paragraphs start may be useful to understand the structure of the content, or to set bookmarks or annotations based on paragraph position.

The options include using double space for softer paragraph breaks, or alternatively the paragraph symbol, Pilcrow (¶), for more obvious paragraph breaks.
```

nvdaes committed:
Apply suggestions from code review

Co-authored-by: Sean Budd <seanbudd123@gmail.com>

seanbudd commented on source/config/configFlags.py:
```diff
@@ -251,3 +251,30 @@ def _displayStringLabels(self):
+	# a symbol also known as "paragraph symbol" or "paragraph marker".
+	# This symbol should translate in braille via LibLouis automatically.
+	# If there is a more appropriate character for your locale,
+	# consider overwriting this (e.g. the reversed pilcrow ⁋).
```
Here's a real world example of why this is useful - since several languages have their own symbol for this

```suggestion
	# consider overwriting this (e.g. for Ge'ez ፨).
```

https://en.wikipedia.org/wiki/Pilcrow#Paragraph_signs_in_non-Latin_writing_systems

nvdaes committed:
Apply suggestions from code review

Co-authored-by: Sean Budd <seanbudd123@gmail.com>

nvdaes committed:
Update source/config/configSpec.py

Co-authored-by: Sean Budd <seanbudd123@gmail.com>

nvdaes commented:
@seanbudd , many thanks for explaining the reason to make the option about paragraph markers an integer. I didn't realise that s.making pilcrows translatable makes this necessary.
I'll finish and test this properly today, in a few hours.

nvdaes committed:
Use paragraph start marker with integer value in config

pre-commit-ci[bot] committed:
Pre-commit autofix

nvdaes committed:
Fix for paragraphStartMarker

pre-commit-ci[bot] committed:
Pre-commit autofix

nvdaes committed:
Merge branch 'master' into brailleParagraphs

nvdaes committed:
Fix paragraphStartMarker in function

nvdaes committed:
Make PILCROW translatable adding a getParagraphStartMarker function in braille.py

nvdaes commented:
@seanbudd , seems that we can not assign a value to PILCROW in the enum class using pgettext or _(). So I've created a new function in braille.py to get the text corresponding to the paragraphStartMarker, and from this functon this can be made translatable.

seanbudd commented on source/config/configFlags.py:
```diff
@@ -256,15 +256,25 @@ def _displayStringLabels(self):
 class ParagraphStartMarker(DisplayStringStrEnum):
 	NONE = ""
 	SPACE = " "
+	# Translators: This is a paragraph start marker used in braille.
+	# The default symbol is the pilcrow,
+	# a symbol also known as "paragraph symbol" or "paragraph marker".
+	# This symbol should translate in braille via LibLouis automatically.
+	# If there is a more appropriate character for your locale,
+	# consider overwriting this (e.g. for Ge'ez ፨).
+	# You can also use Unicode Braille such as ⠘⠏.
+	# Ensure this is consistent with other strings with the context "paragraphMarker".
```
this comment should be deleted now that it has been moved to getParagraphStartMarker

seanbudd commented on source/config/configSpec.py:
```diff
@@ -85,7 +85,7 @@
 		optionsEnum="ReviewRoutingMovesSystemCaretFlag", behaviorOfDefault="NEVER")
 	readByParagraph = boolean(default=false)
-	paragraphStartMarker = option("", " ", "¶", default="")
+	paragraphStartMarker = integer(min=0, max=2, default=0)
```
this change is no longer necessary if the pilcrow is not translatable

seanbudd commented on source/braille.py:
```diff
@@ -1198,6 +1199,26 @@ def getFormatFieldBraille(field, fieldCache, isAtStart, formatConfig):
 	return TEXT_SEPARATOR.join([x for x in textList if x])
 
 
+def getParagraphStartMarker() -> Optional[str]:
```
```suggestion
def getParagraphStartMarker() -> str | None:
```

nvdaes committed:
Update source/braille.py

Co-authored-by: Sean Budd <seanbudd123@gmail.com>

nvdaes committed:
Apply review suggestions

nvdaes committed:
Merge remote-tracking branch 'origin/master' into brailleParagraphs

pre-commit-ci[bot] committed:
Pre-commit auto-fix

nvdaes commented:
pre-commit.ci run

pre-commit-ci[bot] committed:
Pre-commit auto-fix

nvdaes commented:
@seanbudd , I've addressed your comments. Please let me know if anything else needs to be done.
